tensorflow_probability/python/experimental/mcmc/expectations_reducer.py

```diff
@@ -0,0 +1,227 @@
+# Copyright 2020 The TensorFlow Probability Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""ExpectationsReducer for pre-packaged streaming expectations."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+
+# Dependency imports
+import tensorflow.compat.v2 as tf
+
+from tensorflow_probability.python.experimental.mcmc import reducer as reducer_base
+from tensorflow_probability.python.internal import nest_util
+from tensorflow_probability.python.internal import prefer_static as ps
+from tensorflow_probability.python.mcmc.internal import util as mcmc_util
+from tensorflow.python.util import nest  # pylint: disable=g-direct-tensorflow-import
+
+
+__all__ = [
+    'ExpectationsReducer',
+]
+
+
+ExpectationsReducerState = collections.namedtuple(
+    'ExpectationsReducerState', 'num_samples, running_total')
+
+
+class ExpectationsReducer(reducer_base.Reducer):
+  """`Reducer` that computes a running expectation.
+
+  `ExpectationsReducer` calculates expectation over some arbitrary structure
+  of `callables`. A callable is a function that accepts a Markov chain sample
+  and kernel results, and outputs the relevant value for expectation
+  calculation. In other words, if we denote a callable as f(x,y),
+  `ExpectationsReducer` computes E[f(x,y)] for all provided functions. The
+  finalized expectation will also identically mimic the structure of
+  `callables`.
+
+  As with all reducers, ExpectationsReducer does not hold state information;
+  rather, it stores supplied metadata. Intermediate calculations are held in
+  a state object, which is returned via `initialize` and `one_step` method
+  calls.
+  """
+
+  def __init__(self, callables=lambda sample, kr: sample, name=None):
```

Comment:
lambdas as default arguments don't render well in docstrings in, e.g., Colab notebooks. A style we've adopted elsewhere is to move this out to a freestanding function with a descriptive name, e.g.:

```python
def _get_sample(sample, kr):
  return sample
```

You can see the difference if you try this in IPython:

```python
def foo():
    return 0
    
def baz(a=lambda: 0):
    return a()
    
def bar(a=foo):
    return a()
    
help(bar) # Arguably this has more information
help(baz)
```